src/Microsoft.OData.Core/ODataWriterCore.cs

```diff
@@ -1800,15 +1800,22 @@ private void WriteEntityReferenceLinkImplementation(ODataEntityReferenceLink ent
 
             this.CheckForNestedResourceInfoWithContent(ODataPayloadKind.EntityReferenceLink, null);
             Debug.Assert(
-                this.CurrentScope.Item is ODataNestedResourceInfo,
+                this.CurrentScope.Item is ODataNestedResourceInfo || this.ParentNestedResourceInfoScope.Item is ODataNestedResourceInfo,
                 "The CheckForNestedResourceInfoWithContent should have verified that entity reference link can only be written inside a nested resource info.");
 
             if (!this.SkipWriting)
             {
                 this.InterceptException(() =>
                 {
                     WriterValidationUtils.ValidateEntityReferenceLink(entityReferenceLink);
-                    this.WriteEntityReferenceInNavigationLinkContent((ODataNestedResourceInfo)this.CurrentScope.Item, entityReferenceLink);
+
+                    ODataNestedResourceInfo nestedInfo = this.CurrentScope.Item as ODataNestedResourceInfo;
+                    if (nestedInfo == null)
+                    {
+                        nestedInfo = this.ParentNestedResourceInfoScope.Item as ODataNestedResourceInfo;

```

Comment:
>ParentNestedResourceInfoScope [](start = 42, length = 29)

Can we safely assert that ParentNestedResourceInfoScope will always be non-null?  What if the user tries to call WriteEntityReferenceLinkImplementation at the top level?

Reply:
I think we can.

---
In reply to: [289887974](https://github.com/OData/odata.net/pull/1464#discussion_r289887974) [](ancestors = 289887974)